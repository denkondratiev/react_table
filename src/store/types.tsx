import {
  setParams,
  setTable,
  setRows,
  setCells,
  setShowButtons,
  increment,
  setNewRow,
  removeRow
} from './actions'

export const ACTION_TYPES = {
  SET_PARAMS: 'SET::PARAMS',
  SET_TABLE: 'SET::TABLE',
  SET_ROWS: 'SET::ROWS',
  SET_CELLS: 'SET::CELLS',
  SHOW_BUTTONS: 'SHOW::BUTTONS',
  INCREMENT: 'INCREMENT',
  ADD_ROW: 'ADD::ROW',
  REMOVE_ROW: 'REMOVE::ROW'
}

export type State = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
  params: Params
  buttons: boolean
}

export type Action = (
  | ReturnType<typeof setParams>
  | ReturnType<typeof setTable>
  | ReturnType<typeof setRows>
  | ReturnType<typeof setCells>
  | ReturnType<typeof setShowButtons>
  | ReturnType<typeof increment>
  | ReturnType<typeof setNewRow>
  | ReturnType<typeof removeRow>
)

export type Params = {
  [name: string]: string | number
}

export type RowsParams = {
  [name: string]: Array<string>
}

export type CellsParams = {
  [name: string]: { id: string, amount: number }
}

export type NewRowsParams = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
}
