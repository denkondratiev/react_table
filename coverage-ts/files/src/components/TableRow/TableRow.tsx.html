
  <!DOCTYPE html>
  <html>
    <head>
      <title>TableRow.tsx</title>
      <link href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css" type="text/css" rel="stylesheet">
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.52.2/codemirror.min.js" type="text/javascript" charset="utf-8"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.52.2/mode/javascript/javascript.min.js" type="text/javascript" charset="utf-8"></script>
<link href="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.52.2/codemirror.min.css" type="text/css" rel="stylesheet">
<script src="..\..\..\..\assets\source-file.js" type="text/javascript" charset="utf-8"></script>
<link href="..\..\..\..\assets\source-file.css" type="text/css" rel="stylesheet">
    </head>
    <body>
    <div style="margin-top:3em" class="ui container"><h1 class="ui header"><a href="..\..\..\..\index.html">TypeScript coverage report</a></h1><table style="margin-top:2em" class="ui celled table"><thead class=""><tr class=""><th class="">Filename</th><th class="">Percent</th><th class="">Threshold</th><th class="">Total</th><th class="">Covered</th><th class="">Uncovered</th></tr></thead><tbody class=""><tr class="positive"><td class="">src\components\TableRow\TableRow.tsx</td><td class="">100.00%</td><td class="">80%</td><td class="">105</td><td class="">105</td><td class="">0</td></tr></tbody></table><textarea id="editor" readonly="" style="margin-top:3em">import React, { useState } from &#x27;react&#x27;
import { TableRowProps } from &#x27;../../helpers/interface&#x27;
import TableCell from &#x27;../TableCell/TableCell&#x27;
import TableCellSum from &#x27;../TableCellSum/TableCellSum&#x27;

const areEqual = (prevProps: TableRowProps, nextProps: TableRowProps): boolean =&gt; {
  const { row, cells, lightList } = prevProps

  const equal = !row.some((id) =&gt; (
    nextProps.cells[id].amount !== cells[id].amount ||
    nextProps.lightList[id] !== lightList[id]
  ))

  return equal
}

const TableRow: React.FC&lt;TableRowProps&gt; = (props) =&gt; {
  const {
    id,
    row,
    cells,
    rowSum,
    lightList,
    onMouseEnterHandler,
    onMouseLeaveHandler,
    onClickIncrement
  } = props

  const [showPercent, setShowPercent] = useState(false)

  const onMouseEnterPercent = () =&gt; {
    setShowPercent(true)
  }

  const onMouseLeavePercent = () =&gt; {
    setShowPercent(false)
  }

  return (
    &lt;tr
      id={id}
    &gt;
      {row.map(cellId =&gt; {
        const value = cells[cellId].amount
        const percent = (cells[cellId].amount / rowSum * 100).toFixed(2)
        let styleString = &#x27;&#x27;

        if (lightList[cellId]) { styleString += &#x27;#6c757d&#x27; }

        if (showPercent) {
          styleString += `linear-gradient(90deg, rgba(220,53,69,1) ${percent}%, rgba(108,117,125,1) ${percent}%)`
        }

        return (
          &lt;TableCell
            key={cellId}
            id={cellId}
            amount={showPercent ? percent : value}
            styleString={styleString}
            onMouseEnterHandler={onMouseEnterHandler}
            onMouseLeaveHandler={onMouseLeaveHandler}
            onClickIncrement={onClickIncrement}
          /&gt;
        )
      })}
      &lt;TableCellSum
        rowSum={rowSum}
        onMouseEnterPercent={onMouseEnterPercent}
        onMouseLeavePercent={onMouseLeavePercent}
      /&gt;
    &lt;/tr&gt;
  )
}

export default React.memo(TableRow, areEqual)
</textarea><pre id="annotations" style="display:none">[]</pre></div>
    <p class="footer-text">TypeScript Coverage Report generated by <a href="https://github.com/plantain-00/type-coverage">type-coverage</a> and <a href="https://github.com/alexcanessa/typescript-coverage-report">typescript-coverage-report</a> at Thu, 30 Jul 2020 09:48:47 GMT</p>
    </body>
  </html>
  