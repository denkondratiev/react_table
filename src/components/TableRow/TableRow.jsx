import React from 'react'
import { v4 as uuidv4 } from 'uuid'
import { TableRowShape } from '../../helpers/shapes'
import TableCell from '../TableCell/TableCell'
import TableSumCell from '../TableSumCell/TableSumCell'

export const TableRow = (props) => {
  const { row, rowIndex } = props

  let rowSum = null

  return (
    <tr
      key={uuidv4}
      id={rowIndex}
    >
      {row.map((cell, cellIndex) => {
        rowSum = row.map(cell => cell.amount).reduce((prev, cur) => prev + cur, 0)
        return (
          <TableCell
            key={uuidv4()}
            amount={cell.amount}
            id={cell.id}
            cellIndex={cellIndex}
            rowIndex={rowIndex}
            isLight={cell.isLight}
            showPercent={cell.showPercent}
            percent={(cell.amount / rowSum * 100).toFixed(2)}
          />

        )
      })}
      <TableSumCell
        key={uuidv4()}
        rowIndex={rowIndex}
        rowSum={rowSum}
      />
    </tr>
  )
}

TableRow.propTypes = TableRowShape.isRequired
