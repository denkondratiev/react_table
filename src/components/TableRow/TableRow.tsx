import React, { useState } from 'react'
import { TableRowProps } from '../../helpers/interface'
import TableCell from '../TableCell/TableCell'
import TableCellSum from '../TableCellSum/TableCellSum'

const areEqual = (prevProps: TableRowProps, nextProps: TableRowProps): boolean => {
  const { row, cells, lightList } = prevProps

  const equal = !row.some((id) => (
    nextProps.cells[id].amount !== cells[id].amount ||
    nextProps.lightList[id] !== lightList[id]
  ))

  return equal
}

const TableRow: React.FC<TableRowProps> = (props) => {
  const {
    id,
    row,
    cells,
    rowSum,
    lightList,
    onMouseEnterHandler,
    onMouseLeaveHandler,
    onClickIncrement
  } = props

  const [showPercent, setShowPercent] = useState(false)

  const onMouseEnterPercent = () => {
    setShowPercent(true)
  }

  const onMouseLeavePercent = () => {
    setShowPercent(false)
  }

  return (
    <tr
      id={id}
    >
      {row.map(cellId => {
        const value = cells[cellId].amount
        const percent = (cells[cellId].amount / rowSum * 100).toFixed(2)
        let styleString = ''

        if (lightList[cellId]) { styleString += '#6c757d' }

        if (showPercent) {
          styleString += `linear-gradient(90deg, rgba(220,53,69,1) ${percent}%, rgba(108,117,125,1) ${percent}%)`
        }

        return (
          <TableCell
            key={cellId}
            id={cellId}
            amount={showPercent ? percent : value}
            styleString={styleString}
            onMouseEnterHandler={onMouseEnterHandler}
            onMouseLeaveHandler={onMouseLeaveHandler}
            onClickIncrement={onClickIncrement}
          />
        )
      })}
      <TableCellSum
        rowSum={rowSum}
        onMouseEnterPercent={onMouseEnterPercent}
        onMouseLeavePercent={onMouseLeavePercent}
      />
    </tr>
  )
}

export default React.memo(TableRow, areEqual)
