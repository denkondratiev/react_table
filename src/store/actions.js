export const ACTION_TYPES = {
  SET_COLUMNS: 'SET::COLUMNS',
  SET_ROWS: 'SET::ROWS',
  SET_TABLE: 'SET::TABLE',
  SET_HIGHLIGHT: 'SET::HIGHLIGHT',
  SET_SHOW_TABLE: 'SET::SHOW::TABLE',
  SET_SHOW_BUTTONS: 'SET::SHOW::BUTTONS'
}

export const setColumns = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_COLUMNS
})

export const setRows = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_ROWS
})

export const setHighlights = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_HIGHLIGHT
})

export const setTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_TABLE
})

export const setShowTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_TABLE
})

export const setShowButtons = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_BUTTONS
})
