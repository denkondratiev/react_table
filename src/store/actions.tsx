import {
  ACTION_TYPES,
  Params,
  RowsParams,
  CellsParams,
  NewRowsParams
} from './types'

export const setParams = (params: Params) => ({
  payload: params,
  type: ACTION_TYPES.SET_PARAMS
})

export const setTable = (params: Array<string>) => ({
  payload: params,
  type: ACTION_TYPES.SET_TABLE
})

export const setRows = (params: RowsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_ROWS
})

export const setCells = (params: CellsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_CELLS
})

export const setShowButtons = (params: boolean) => ({
  payload: params,
  type: ACTION_TYPES.SHOW_BUTTONS
})

export const increment = (params: string) => ({
  payload: params,
  type: ACTION_TYPES.INCREMENT
})

export const setNewRow = (params: NewRowsParams) => ({
  payload: params,
  type: ACTION_TYPES.ADD_ROW
})

export const removeRow = (params: any) => ({
  payload: params,
  type: ACTION_TYPES.REMOVE_ROW
})
