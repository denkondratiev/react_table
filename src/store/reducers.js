import { createStore } from 'redux'
import { ACTION_TYPES } from './actions'

function rootReducer (state, action) {
  const {
    SET_COLUMNS,
    SET_ROWS,
    SET_HIGHLIGHTS,
    SET_TABLE,
    SET_SHOW_TABLE,
    SET_SHOW_BUTTONS,
    INCREMENT,
    SET_LIGHT_CELLS
  } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE: {
      return {
        ...state,
        table: action.payload
      }
    }

    case SET_COLUMNS: {
      return {
        ...state,
        columns: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_ROWS: {
      return {
        ...state,
        rows: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_HIGHLIGHTS: {
      return {
        ...state,
        highlights: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_SHOW_TABLE: {
      return {
        ...state,
        showTable: action.payload
      }
    }

    case SET_SHOW_BUTTONS: {
      return {
        ...state,
        showButtons: action.payload
      }
    }

    case INCREMENT: {
      state.table.splice(action.payload.id, 1, [...state.table[action.payload.id]
        .map((item, index) => index !== action.payload.cellIndex
          ? item
          : {
            ...state.table[action.payload.id][action.payload.cellIndex],
            amount: state.table[action.payload.id][action.payload.cellIndex].amount + 1
          })
      ])
      const newTable = [...state.table]
      return {
        ...state,
        table: newTable
      }
    }

    case SET_LIGHT_CELLS: {
      // const amount = action.payload.amount
      // const highlights = action.payload.highlights
      const rowId = action.payload.id
      const cellIndex = action.payload.cellIndex
      // const col = state.table.flat().length / state.table.length
      console.log(rowId, cellIndex)
      break

      // const lightArray = [...state.table].flat().map((item, index) => ({
      //   ...item,
      //   rowIndex: index,
      //   difference: Math.abs(item.amount - amount)
      // })).sort((a, b) => a.difference - b.difference).slice(0, highlights + 1)

      // state.table.map(cell => (
      //   lightArray.some(lightCell => lightCell.cellId === cell.cellId)
      //     ? { ...cell, isLight: true }
      //     : cell
      // ))

      // return {
      //   ...state,
      //   table: newTable
      // }
    }

    default:
      return state
  }
}

const initialState = {
  rows: '',
  columns: '',
  highlights: '',
  table: [],
  showTable: false,
  showButtons: false
}

const store = createStore(
  rootReducer,
  initialState,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

export default store
