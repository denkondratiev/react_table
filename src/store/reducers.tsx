import { combineReducers } from 'redux'
import { ACTION_TYPES } from './actions'
import { Action, Params, RowsParams, CellsParams } from './actions'

const paramsReducer = (state: Params = {}, action: Action): Params => {
  const { SET_PARAMS } = ACTION_TYPES

  switch (action.type) {
    case SET_PARAMS:
      return action.payload
    default:
      return state
  }
}

function tableReducer (state: Array<string> = [], action: Action): Array<string> {
  const { SET_TABLE, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE:
      return action.payload
    case ADD_ROW:
      return [...state, action.payload.table]
    case REMOVE_ROW:
      return state.filter(id => id !== action.payload.lastRow)
    default:
      return state
  }
}

const rowsReducer = (state: RowsParams = {}, action: Action): RowsParams => {
  const { SET_ROWS, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_ROWS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.rows }
    case REMOVE_ROW:
      delete state[action.payload.lastRow]
      return state
    default:
      return state
  }
}


const cellsReducer = (state: CellsParams = {}, action: Action): CellsParams => {
  const { SET_CELLS, INCREMENT, ADD_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_CELLS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.cells }
    case INCREMENT:
      return {
        ...state,
        [action.payload]: {
          ...state[action.payload],
          amount: state[action.payload].amount + 1
        }
      }
    default:
      return state
  }
}

const buttonsReducer = (state: boolean = false, action: Action): boolean => {
  const { SHOW_BUTTONS } = ACTION_TYPES

  switch (action.type) {
    case SHOW_BUTTONS:
      return action.payload
    default:
      return state
  }
}

export const rootReducer = combineReducers({
  params: paramsReducer,
  table: tableReducer,
  rows: rowsReducer,
  cells: cellsReducer,
  buttons: buttonsReducer
})
