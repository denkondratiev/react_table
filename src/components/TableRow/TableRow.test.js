import React from 'react'
import { Provider } from 'react-redux'
import TableRow from './TableRow'
import TestRenderer from 'react-test-renderer'
import { createStore } from 'redux'
import { rootReducer } from '../../store/reducers'

it('should correct render table row with some value', () => {
  const params = { rowsAmount: 2, columnsAmount: 2, lightsAmount: 2 }
  const table = ['row1', 'row2']
  const rows = {
    row1: ['somecell1', 'somecell2'],
    row2: ['somecell3', 'somecell4']
  }

  const cells = {
    somecell1: { id: 'somecell1', amount: 1 },
    somecell2: { id: 'somecell2', amount: 2 },
    somecell3: { id: 'somecell3', amount: 0 },
    somecell4: { id: 'somecell4', amount: 1 }
  }
  const buttons = false
  const lightList = ['somecell1', 'somecell2']
  const store = createStore(rootReducer, { params, table, rows, cells, buttons })

  const tree = TestRenderer
    .create(
      <Provider store={store}>
        <TableRow
          row={rows.row1}
          rowSum={3}
          cells={cells}
          lightList={lightList}
        />
      </Provider>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
