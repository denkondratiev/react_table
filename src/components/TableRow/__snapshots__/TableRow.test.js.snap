// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correct render table row with some value 1`] = `
<tr>
  <td
    id="somecell1"
    style={
      Object {
        "background": "",
      }
    }
  >
    1
  </td>
  <td
    id="somecell2"
    style={
      Object {
        "background": "",
      }
    }
  >
    2
  </td>
  <td
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Sum: 
    3
  </td>
</tr>
`;
